### PR TITLE
[WIP]  Output clickhouse

### PR DESCRIPTION
**Pull request by pachico**
Hi, this is still work in progress (features, docs and tests still missing) but basic functionality is there.
Do you think you can have a look at it and let me know if I'm totally misguided or if you see this as a possible merge in the future?

Thanks in advance

### CODE REVIEW

File: output/clickhouse/clickhouse.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+	newEventAsJSON, err := jsoniter.Marshal(newEventMap)
+
+	if err != nil {
+		return err
+	}
+
+	queryString := fmt.Sprintf("INSERT INTO %s FORMAT JSONEachRow", t.TargetTable)
+	requestURL := fmt.Sprintf("%s?query=%s", t.DSN, url.QueryEscape(queryString))
+
+	request, _ := http.NewRequest("POST", requestURL, strings.NewReader(string(newEventAsJSON)))
```
Comment (tsaikd) [inline output/clickhouse/clickhouse.go:101]: Please do not ignore any errors.

File: output/clickhouse/clickhouse.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+// @todo: create bulk functionality
+// @todo: add gzip functionality
+func (t *OutputConfig) Output(ctx context.Context, event logevent.LogEvent) (err error) {
+
+	newEventMap := make(map[string]interface{})
+
+	for _, key := range t.Fields {
+		fieldValue, exists := event.GetValue(key)
+
+		if true == exists {
```
Comment (tsaikd) [inline output/clickhouse/clickhouse.go:77]:
```suggestion
		if exists {
```

File: output/clickhouse/clickhouse.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
+package outputclickhouse
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"log"
```
Comment (tsaikd) [inline output/clickhouse/clickhouse.go:7]: try `github.com/tsaikd/gogstash/config/goglog`

File: output/clickhouse/clickhouse.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+	request, _ := http.NewRequest("POST", requestURL, strings.NewReader(string(newEventAsJSON)))
+	resp, err := t.httpClient.Do(request)
+
+	if err != nil {
+		log.Println("Error executing request to Clickhouse: ", err)
+	}
+
+	if resp.StatusCode != 200 {
+		defer resp.Body.Close()
```
Comment (tsaikd) [inline output/clickhouse/clickhouse.go:109]: https://godoc.org/net/http
> The client must close the response body when finished with it

Does the `pester` HTTP client has different behavior?

File: output/clickhouse/clickhouse.go (hunk 1 above)
Comment (pachico) [reply output/clickhouse/clickhouse.go:101]: Sure thing, will do.

File: output/clickhouse/clickhouse.go (hunk 3 above)
Comment (pachico) [reply output/clickhouse/clickhouse.go:7]: I see not all plugins use it but if it's the desired one I will implement it.

File: output/clickhouse/clickhouse.go (hunk 4 above)
Comment (pachico) [reply output/clickhouse/clickhouse.go:109]: Yes, and I want to implement it. It mostly handles retries, which reduces the possibility of missed documents. I want users to be able to customise it as they do for elastic plugin.

**Comment by pachico** [comment]
I addition, I want to implement batch processing since ClickHouse prefers it

**Comment by dairongpeng** [comment]
Hi, is there a plan to merge?

**Comment by tsaikd** [comment]
I don't see any actions after the last review.

**Comment by pachico** [comment]
I won't proceed with this draft since I'm currently using another solution that already has ClickHouse output option.

**Pull request closed without merging**